[QUIC] API QuicConnection

ManickaP opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/68902
Contributes to https://github.com/dotnet/runtime/issues/69675

~This PR is based on https://github.com/dotnet/runtime/pull/71579 so it shows more changes that it actually contains.~

ManickaP committed:
QuicConnection new API including compilable implementation

ManickaP committed:
Fixed logging

ManickaP committed:
Fixed S.N.Quic and S.N.Http tests

ManickaP committed:
Options now correspond to the issue

ManickaP committed:
Feedback

ManickaP committed:
Feedback

ManickaP committed:
Listener comment; PreviewFeature attribute

ManickaP committed:
Comments, PreviewFeature attribute and RemoteCertificate disposal.

ManickaP committed:
Preview feature attribute is assembly wide

ManickaP committed:
Some typos

ManickaP committed:
Fixed test with certificate

ManickaP committed:
Default values as constants

ManickaP committed:
Event handlers split into methods called via switch expression.

ManickaP committed:
Some more comments

ManickaP committed:
Unified unsafe usage

ManickaP committed:
Fixed some more tests

ManickaP committed:
Cleaned up some exceptions and resource strings.

ManickaP commented:
This is fairly stable now, ready for review. Still letting it in draft due to unapproved API.
@CarnaViire @wfurt @rzikm if you could give it a look.

Also note that stream will be soon in PR as well.

CarnaViire commented on src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs:
```diff
@@ -1,93 +1,597 @@
+        {
+            if (NetEventSource.Log.IsEnabled())
+            {
+                NetEventSource.Error(null, $"Received event {connectionEvent->Type}");
```
Should we elaborate in the message? and explain that it's not allocated anymore, or it's some strange instance (how could that be??)

ManickaP replied:
Fixed all occurrences, also fixed in stream.

ManickaP commented on src/libraries/System.Net.Quic/tests/FunctionalTests/QuicStreamTests.cs:
```diff
@@ -814,13 +811,15 @@ public async Task WriteAsync_LocalAbort_Throws()
         public async Task WaitForWriteCompletionAsync_ServerWriteAborted_Throws()
         {
             const int ExpectedErrorCode = 0xfffffff;
+            SemaphoreSlim sem = new SemaphoreSlim(0);
```
These 3 lines are not strictly needed for this PR and I'll remove them. But they need to be introduced in stream PR, where we're getting rid of `ShutdownCompleted` (we could replace it with `await Task.WhenAll(ReadsClosed, WritesClosed)` or something similar I guess, I haven't thought about that when fixing the tests), since then you're left with `DisposeAsync` which will attempt to abort read side which will race with `AbortWrite` from the server and sometimes win. Or the semaphore could be replaced with await stream.ReadsClosed on the client side and check for StreamAborted exception.

ManickaP commented on src/libraries/System.Net.Quic/src/System/Net/Quic/QuicListener.cs:
```diff
@@ -12,6 +10,8 @@
 using System.Threading.Channels;
 using System.Threading.Tasks;
 using Microsoft.Quic;
+using static System.Net.Quic.MsQuicHelpers;
+using static System.Net.Quic.QuicDefaults;
```
I've used this mainly for being able to use MsQuic status constants without MsQuic, e.g.: `QUIC_STATUS_HANDSHAKE_FAILURE` vs `MsQuic.QUIC_STATUS_HANDSHAKE_FAILURE`. The status constants contain `QUIC_` in their name which is self-explanatory.
I'm not opposed to removing the `using static` or limiting the usage only to the status constants. After all this is a cosmetic change I don't have a strong preference about. I just want to stay consistent with its usage across the S.N.Quic.

ManickaP replied:
I've removed all usages that didn't involve msquic and its status codes.

ManickaP committed:
Feedback

wfurt commented on src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicHelpers.cs:
```diff
@@ -0,0 +1,82 @@
+
+    internal static unsafe QuicAddr ToQuicAddr(this IPEndPoint iPEndPoint)
+    {
+        // TODO: is the layout same for SocketAddress.Buffer and QuicAddr on all platforms?
```
it should AFAIK. It would be nice to have some assert about it. It differs among platforms but MsQUIC now uses native layout e.g. same as SocketAddress.

ManickaP replied:
https://github.com/dotnet/runtime/issues/72027

ManickaP committed:
Merge branch 'main' into mapichov/quic-connection

ManickaP committed:
Latest greatest API proposal.

ManickaP commented:
This is update based on the last version of API from https://github.com/dotnet/runtime/issues/68902#issuecomment-1182102876
Main changes:
- behavior of DisposeAsync (automatically will close with DefaultCloseErrorCode)
- removed required from the options and replaced by explicit validation

@CarnaViire @wfurt @rzikm please give it look

ManickaP committed:
Fixed Http solution

rzikm commented on src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs:
```diff
@@ -1,93 +1,604 @@
+
+            if (!options.RemoteEndPoint.TryParse(out string? host, out IPAddress? address, out int port))
+            {
+                throw new ArgumentException($"Unsupported remote endpoint type '{options.RemoteEndPoint.GetType()}', expected IP or DNS endpoint.", nameof(options));
```
This should go to localizable string resources.

ManickaP replied:
Fixed

ManickaP committed:
Merge branch 'main' into mapichov/quic-connection

rzikm commented on src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs:
```diff
@@ -1,93 +1,604 @@
+    /// </summary>
+    /// <remarks>
+    /// Connection close is not graceful in regards to its streams, i.e.: calling <see cref="CloseAsync(long, CancellationToken)"/> will immediately abort all streams associated with this connection.
+    /// Please make sure, that all streams have been closed and all their data consumed before calling this method;
```
I don't think we should use the word "please" in comments. But I think we will need our docs team to proofread the comments anyway...

ManickaP replied:
Done

rzikm commented on src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnectionOptions.cs:
```diff
@@ -20,48 +21,110 @@ internal QuicConnectionOptions()
+    /// <param name="argumentName">Name of the from the caller.</param>
+    internal virtual void Validate(string argumentName)
+    {
+        if (MaxInboundUnidirectionalStreams < 0 || MaxInboundBidirectionalStreams > ushort.MaxValue)
+        {
+            throw new ArgumentOutOfRangeException($"{nameof(QuicConnectionOptions.MaxInboundBidirectionalStreams)} should be within [0, {ushort.Max}) range.", argumentName);
```
Is this limitation of MsQuic? I think the RFC does not impose any limits on concurrently opened streams.

ManickaP replied:
```
uint16_t PeerBidiStreamCount;
uint16_t PeerUnidiStreamCount;
```
https://github.com/microsoft/msquic/blob/main/docs/api/QUIC_SETTINGS.md

rzikm commented on src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnectionOptions.cs:
```diff
@@ -20,48 +21,110 @@ internal QuicConnectionOptions()
+    public QuicClientConnectionOptions()
+    {
+        MaxInboundBidirectionalStreams = QuicDefaults.DefaultClientMaxInboundBidirectionalStreams;
+        MaxInboundUnidirectionalStreams = QuicDefaults.DefaultServerMaxInboundUnidirectionalStreams;
```
Possible copy-paste error?

```suggestion
        MaxInboundUnidirectionalStreams = QuicDefaults.DefaultClientMaxInboundUnidirectionalStreams;
```

ManickaP replied:
Fixed

rzikm commented on src/libraries/System.Net.Quic/src/System/Net/Quic/QuicDefaults.cs:
```diff
@@ -0,0 +1,35 @@
+    /// <summary>
+    /// Max value for application error codes that can be sent by QUIC, see <see href="https://www.rfc-editor.org/rfc/rfc9000.html#integer-encoding"/>.
+    /// </summary>
+    public const long MaxErrorCodeValue = 1L << 62;
```
this is 63 bit value, last valid value is `(1L << 62) - 1`

ManickaP replied:
Good catch

ManickaP replied:
Fixed

rzikm reviewed (approved):
LGTM, modulo comments.

ManickaP committed:
Feedback

ManickaP commented:
Failure is https://github.com/dotnet/runtime/issues/71890, it's been closed in the meantime.

ManickaP commented:
/backport to release/7.0-preview7

ManickaP commented:
@JamesNK this will need your reaction as well. Note that the options classes were changed in the last API review and we removed the `required` keyword. If you hit any issues with this PR, let me know.
